Table of content:

1. Variables
2. Functions
3. Strings
4. Lists
5. Dictionary
6. Tuples
7. Booleans
8. Functions again
9. Loops : for
10. If statement
11. While statement
12. Libraries import
13. Numpy
14. Pandas

## Python Tutorial


Welcome to this Python tutorial. The Jupyter Notebook interface is composed of several cells, in which you can write some code. When you run a cell, the only lines executed are those in the cell. However, they are linked by a single kernel : it means that you can define a variable in a cell and use it in an other.

## 1) Variables

To run a cell, select it and press "execute", or shift + enter. Try to run the following cell :

In [1]:
print("Congratulations")

Congratulations


If your code includes some functions which should display some informations, this information will appear under the box, like with "Congratulations" in the previous example.

#### &#x270d; 1.1) Display the number 13 and the message "Python is great"

In [3]:
print(13)
print("Python is great")

13
Python is great


#### &#x270d; 1.2) Create the variables $a = 5$, $b = 5.2$ and $c = 3 + 2j$

In [62]:
a=5
b=5.2
c=3+2j

#### &#x270d; 1.3) Display $a$, $b$ and $c$ with the print function.  
They can be displayed with only one call by separating them with commas in the brackets.

In [63]:
print("a =",a,"\t","b =",b,"\t","c =",c) 

a = 5 	 b = 5.2 	 c = (3+2j)


#### &#x270d; 1.4) Display $a + b$ and $b * c$ without saving the product in a variable.

In [64]:
print("a + b =",a+b)
print("b * c =",b*c)

a + b = 10.2
b * c = (15.600000000000001+10.4j)


#### &#x270d; 1.5) Check the type of $a$, $b$, $c$, then display and show the type of :  
$a + b$  
$a * b$  
$b - c$  
$a + b + c$  
$a * c$  
$c / a$  

In [65]:
print("a est de type :",type(a))
print("b est de type :",type(b))
print("c est de type :",type(c))
print("(a + b) est de type :",type(a + b))
print("(a * b) est de type :",type(a * b))
print("(b - c) est de type :",type(b - c))
print("(a + b + c) est de type :",type(a + b + c))
print("(a * c) est de type :",type(a * c))
print("(c / a) est de type :",type(c / a))

a est de type : <class 'int'>
b est de type : <class 'float'>
c est de type : <class 'complex'>
(a + b) est de type : <class 'float'>
(a * b) est de type : <class 'float'>
(b - c) est de type : <class 'complex'>
(a + b + c) est de type : <class 'complex'>
(a * c) est de type : <class 'complex'>
(c / a) est de type : <class 'complex'>


## 2)Functions

#### &#x270d; 2.1) Save round(b) in a variable e, then display the type of :  
$b$  
$e$  
round

In [66]:
e = round(b)
print("e est de type :",type(e))
print("round est de type :",type(round)) 

e est de type : <class 'int'>
round est de type : <class 'builtin_function_or_method'>


#### &#x270d; 2.2) Create a function which takes as argument a value $x$, and returns $x + x^2$ (show how your function works with example)

In [67]:
def f1(x):
    return x + x**2

f1(2) 

6

#### &#x270d; 2.3) Create a function which takes as argument a value $x$, prints $x$, $2x$, $3x$, and returns $x - 1$. (show how your function works with example)

In [52]:
def f2(x):
    print(x, "|", 2*x, "|", 3*x)
    return x-1

print("function f2 return :", f2(2))

2 | 4 | 6
function f2 return : 1


#### &#x270d; 2.4) Write down a function that takes $x$ and $y$ as the arguments, and returns a rounded value of $x$ multiplied by the value of $y$. (show how your function works with example)

In [68]:
def f3(x,y):
    return round(x*y)

f3(4.3, 2) 

9

## 3) Strings

#### &#x270d; 3.1) Create a string called birthday containing your birthday.

In [54]:
birthday = "June 22"

#### &#x270d; 3.2) Using a and b, create a string "Hello my dear" and print it.

In [70]:
a = "Hello"
b = "my dear"
print(a + " " + b)

Hello my dear


#### &#x270d; 3.3) Print the number of letters of a and b using the function len(string)

In [72]:
print(len(a), len(b))

5 7


A string can be included in the middle of another one using `%s`

Example: 

`animal = "beef"`

`action = "love"`

`dream = "i would %s to eat a %s" %(action,animal)`

`--> dream = "i would love to eat a beef"`

#### &#x270d; 3.4) Create the string "I am 30 years old and I live in Bogota" using the following strings :

In [77]:
age = "30 years old"
city = "Bogota"
print("I am %s and i live in %s"%(age, city))
print(f'I am {age} and I live in {city}')

I am 30 years old and i live in Bogota
I am 30 years old and I live in Bogota


## 4) Lists

To create a list, you need to use `[ ]`. If you're using a mac, the command is alt + shift + `(` or `)` .


#### &#x270d; 4.1) Create a list containing every letters from A to H, called letters.

In [102]:
letters = [chr(i) for i in range(65, 73)] 
letters

['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H']

#### &#x270d; 4.2) Print the letter E using its index in the list letters.

Remember : index of the first element is $0$.

In [104]:
print(letters[4]) 

E


#### &#x270d; 4.3) Replace the letter G by Y in letters.

In [113]:
#letters[-2] = 'Y'
letters[letters.index('G')] = 'Y'

#### &#x270d; 4.4) Print the letter D in two different ways : once with its index, one with its index starting from the last element (negative index).

In [117]:
print(letters[3])
print(letters[-5]) 

D
D


#### &#x270d; 4.5) Add the letter P to the list letters, using the function letters.append()

In [119]:
letters.append('P')

#### &#x270d; 4.6) Create a sublist sublet = ["A","B","C"] from the list letters.

Remember : list[i:j] creates a sublist from list starting from the i th element to the j-1 th element.

In [124]:
sublet = letters[:3] 
sublet

['A', 'B', 'C']

#### &#x270d; 4.7) Concatenate letters and sublet into letters.

In [127]:
letters = letters + sublet
letters

['A', 'B', 'C', 'D', 'E', 'F', 'Y', 'H', 'P', 'A', 'B', 'C', 'A', 'B', 'C']

#### &#x270d; 4.8)  Erase every occurence of C in letters.

Remember : use list.remove(string) to delete every occurence of string in list.

In [134]:
while 'C' in letters: letters.remove('C') 

#### &#x270d; 4.9) Erase the second occurence of A in letters.
Remember : use list.pop(index) to delete the index th element.

In [137]:
letters.pop(-4)

'A'

In [139]:
letters

['A', 'B', 'D', 'E', 'F', 'Y', 'H', 'P', 'B', 'A', 'B']

## 5) Dictionary

Dictionaries are similiar to lists, but instead of accessing an element through its index, we can use some keywords, which can be of any type.
When you create it, you indicate the keyword and then the associated variables.
Try to print every variable of the following dict :

In [141]:
dict = {'x' : 3, 'y' : True, 'z' : [False, 89], 5 : "banana",  True : False }

#### &#x270d; 5.1) A list of the keywords of dict can be accessed using the function `dict.keys()`, use it.

In [178]:
#[dict[i] for i in dict.keys()] 
print("------- USING KEYS --------")
for i in dict.keys():
    print(dict[i]) 
print("\n------- USING ITEMS --------")
for j in dict.items():
    print(dict[j[0]])

------- USING KEYS --------
3
True
[False, 89]
banana
False

------- USING ITEMS --------
3
True
[False, 89]
banana
False


#### &#x270d; 5.2) Add a new value False to dict, indexed by the keyword Banana.


To add a new keyword, give the associated value as if the keyword already existed. Example:  

`dict = {}`

`dict[3] = "hi"`

`--> dict = {3 : "hi"}`

In [181]:
dict["banana"] = False
print(dict)

{'x': 3, 'y': True, 'z': [False, 89], 5: 'banana', True: False, 'banana': False}


#### &#x270d; 5.3) Delete the key `y` from dict.

You can use `dict.pop(key)` to do that.

In [182]:
dict.pop('y')
print(dict)

{'x': 3, 'z': [False, 89], 5: 'banana', True: False, 'banana': False}


#### &#x270d; 5.4) Try to delete the key `g`, which is not in dict.

In [185]:
dict.pop('g') 

KeyError: 'g'

As you can see, instead of just ignoring the command, an error is returned. If you do not want the program to be stopped if the value is not existing, you can add another action to be performed if the key does not exist. It must be written as second parameter.

Addding None as second parameter prevents the program from being stopped by an unexisting key.

#### &#x270d; 5.5) Check this with the previous example.

In [188]:
dict.pop('g',None)  

In [189]:
print(dict) 

{'x': 3, 'z': [False, 89], 5: 'banana', True: False, 'banana': False}


## 6) Tuples
Tuples are very similar to lists, except their shape cannot be changed.

#### &#x270d; 6.1) Create a tuple containing the integers from 5 to 9 and a tuple containing the integers from 1 to 4. Use ( ) to create them.

In [201]:
T2 = (5,6,7,8,9)
T1 = (1,2,3,4) 

#### &#x270d; 6.2) Try to sum up these tuples.

In [202]:
T = T1 + T2
T

(1, 2, 3, 4, 5, 6, 7, 8, 9)

#### &#x270d; 6.3) Print the 5th element of the sum, using two different ways.

In [206]:
print(T[4]) 
print(T[-5]) 

5
5


## 7) Booleans


#### &#x270d; 7.1) Create two booleans `a = True` and `b = False`, and check the result of every combination of `a` and `b` with the logical operators `and` / `or`.

In [207]:
a = True
b = False
OR = a or b
AND = a and b
print(OR)
print(AND) 

True
False


#### &#x270d; 7.2) To invert a boolean, use the function not. Check the result of "not a or b"

In [209]:
not a or b 

False

#### &#x270d; 7.3) Check how to use the different logical operators :  
equals ==  
less than <  
more than >    
less or equal <=
more or equal >=



In [210]:
print(a==b, 3<1, 4>3, 2<=4, 4>=5) 

False False True True False


#### &#x270d; 7.4) Create a the boolean which indicates if 3 is in the list "list" or not.

Use the function in :  `a in list = True` if a is in list and `False` otherwise.

In [214]:
my_list = [4,3.2,9,44,2,"hey",False,True]  
a = 3 in my_list
a

False

## 8) Functions again


Once created, a function is used only if it is called.

Here, the function square returns the squared value of the input.

#### &#x270d; 8.1) Try to print the squared value of 5 using the square function below.

In [216]:
def square(n):
    result = n**2
    return(result)

print(square(5))

25


The "return" word is very important when you define a function, for two reasons. First, it allows to tell what will be the output of the function. Second, this word means that we are at the end of the function, so the next lines do not need to be read.

#### &#x270d; 8.2) Create a function `square2(n)` identical to `square(n)` without the `return(result)` line and see what happens if you try to compute `square2(5)`.

In [227]:
def square2(n):
    x = n*n
    x
    
z = square2(5)
print(z)

None


#### &#x270d; 8.3) Create a function `square3(n)` identical to `square(n)` with one more line `print("hi")` after `return(result)` and see what happens if you try to compute `square3(5)`.

In [229]:
def square3(n):
    return n*n
    print("hi")
    
print(square3(5))

25


#### &#x270d; 8.4) Create a function `firstandlast(l)` which returns a tuple containing the first and the last elements of `l`, a list input.
Try it on the list test.

In [254]:
def firstandlast(l):
    #return l[0], l[-1]
    return tuple(l[::len(l)-1])

L = [1,4,"bbg",False,56,"Ohh !"] 
firstandlast(L)

(1, 'Ohh !')

Once a function is created, it can be called even in the definition of an other function.

#### &#x270d; 8.5) Create a function `sumoffandl` which takes as input a list, and returns the sum of its first and last terms. Use the `firstandlast(n)` function.

In [293]:
def sumoffandl(l):
  first, last = firstandlast(l)
  return first + last

In [295]:
List = ["Good",4,27,[3,28],"animal","X",True, "job !"]
sumoffandl(List) 

'Goodjob !'

## 9) Loops : `for`


Run the following code and watch the syntaxe of the For loop function

In [296]:
my_list = [2,5,4,2,8,20]
for x in my_list:
    print(x)

2
5
4
2
8
20


#### &#x270d; 9.1) Now try your own loop : compute the sum of the terms of the list `my_list` with a for loop.

In [298]:
result=0
for x in my_list:
    result+=x
result    

41

When we want to iterate on some integers, we use the range function : 

`for k in range(i,j)`

 means that `k` will take the value of every integer between `i` and `j-1`

#### &#x270d; 9.2) Compute the product of integers from 37 to 54 using a for loop and the range function.

In [299]:
P=1
for k in range(37,55):
    P = P*k

print(P) 

620558705531887159346749440000


#### &#x270d; 9.3) Create a function which takes a list as input and returns the same list but elements are inverted (the first element becomes the last, etc)

In [6]:
def inverse(l):
    return l[::-1]

def inverse2(l):
    newList = []
    for e in l: 
        # [e] + newList
        newList.insert(0,e)
    return newList
        
Liste = [1,4,8,"vba",56,False,"yes",45]
print(inverse(Liste))
print(inverse2(Liste))

[45, 'yes', False, 56, 'vba', 8, 4, 1]
[45, 'yes', False, 56, 'vba', 8, 4, 1]


## 10) `If` statement

#### &#x270d; 10.1) Test the function test with -5, 1, -1.7, 8, and interpret.

In [34]:
def test(x):
    if x**2 > 4 : #true/false
        print("yes")
    else :
        print("no")
        
test(-5)
test(1)
test(-1.7)
test(8)

yes
no
no
yes


#### &#x270d; 10.2) Create a function which takes a list as input and returns a boolean, set to `True` if the first element is a string, an `False` otherwise.

In [39]:
def firstString(l):
    #return isinstance(l[0], str)
    if type(l[0])==str:
        return True
    return False
    

List2 = ["coucou", False, True, 45, 643, 21]
print(firstString(Liste))
print(firstString(List2))

False
True


#### &#x270d; 10.3) Create a function which takes a list as input and returns a boolean, set to `True` if there is a string in the list, and `False` otherwise.

Hint : combine `for` and `if`.

In [45]:
def hasString(l):
    for e in l:
        if type(e) == str:
            return True
    return False

List3=[1,2,False,3]
List4=[2,4,True,['A'],"dff",32,4]
hasString(Liste), hasString(List3), hasString(List4)

(True, False, True)

## 11) `While` statement

#### &#x270d; 11.1) Run the following cell and interpret.

In [46]:
x = 16
while x > 3 :
    print(x)
    x = x / 2

16
8.0
4.0


#### &#x270d; 11.2) Create a function which takes a list, and returns false if there is no string inside, and the index of the first string otherwise, using a while loop.

In [23]:
def indexString(l):
    i=0
    while( i < len(l) ):
        if(type(l[i]) == str):
            return i
        i+=1
    return False

In [60]:
List6 = [5,4,3,10]
v = indexString(List6)
print(v)

False


#### &#x270d; 11.3) Print the first integer whose squared value is more than 7533493

In [61]:
while v**2 <= 7533493:
    v+=1
print(v) 

2745


## 12) Libraries import

### &#x270d; 12.1) Import the library numpy, with the shortened name np.

Remember : `import XXXXXX as YY` to import the module `XXXXXX` with the shortened name `YY`.

In [62]:
import numpy as np

Once a library is imported, all its functions are stored in the kernel and can be accessed.

#### &#x270d; 12.2) Print the value of pi using the function pi in numpy.

Remember : To use a function of a library imported as YY, use YY.function()

In [63]:
print(np.pi) 

3.141592653589793


## 13) Numpy

#### &#x270d; 13.1) Create an array containing the integers from 0 to 9

In [66]:
A = np.arange(10)
A

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

#### &#x270d; 13.2) Create a 10x10 matrix name M, use arange and reshape

In [69]:
M = np.arange(25,125).reshape(10,10) 
M

array([[ 25,  26,  27,  28,  29,  30,  31,  32,  33,  34],
       [ 35,  36,  37,  38,  39,  40,  41,  42,  43,  44],
       [ 45,  46,  47,  48,  49,  50,  51,  52,  53,  54],
       [ 55,  56,  57,  58,  59,  60,  61,  62,  63,  64],
       [ 65,  66,  67,  68,  69,  70,  71,  72,  73,  74],
       [ 75,  76,  77,  78,  79,  80,  81,  82,  83,  84],
       [ 85,  86,  87,  88,  89,  90,  91,  92,  93,  94],
       [ 95,  96,  97,  98,  99, 100, 101, 102, 103, 104],
       [105, 106, 107, 108, 109, 110, 111, 112, 113, 114],
       [115, 116, 117, 118, 119, 120, 121, 122, 123, 124]])

#### &#x270d; 13.3) Compute the eigenvalues and eigenvectors of M, using the function linalg.eig from numpy.

In [87]:
val, vect = np.linalg.eig(M)
print(val)
print(vect) 

[ 7.55913941e+02+0.00000000e+00j -1.09139408e+01+0.00000000e+00j
 -2.48345445e-14+1.33275088e-14j -2.48345445e-14-1.33275088e-14j
  9.08095664e-16+5.42945679e-15j  9.08095664e-16-5.42945679e-15j
 -3.61806583e-15+0.00000000e+00j  1.17575418e-15+1.61828934e-15j
  1.17575418e-15-1.61828934e-15j -5.43427133e-16+0.00000000e+00j]
[[-0.11963042+0.j          0.50708361+0.j          0.08402386-0.13707205j
   0.08402386+0.13707205j -0.026842  -0.04906535j -0.026842  +0.04906535j
   0.03534995+0.j          0.00606806-0.00237097j  0.00606806+0.00237097j
   0.02150041+0.j        ]
 [-0.15873633+0.j          0.39706663+0.j         -0.41396302+0.16212135j
  -0.41396302-0.16212135j -0.0246189 +0.06070156j -0.0246189 -0.06070156j
  -0.06627864+0.j          0.01804167+0.00906418j  0.01804167-0.00906418j
   0.00450502+0.j        ]
 [-0.19784224+0.j          0.28704965+0.j          0.09930343+0.01757413j
   0.09930343-0.01757413j -0.13486837+0.0769372j  -0.13486837-0.0769372j
   0.24826249+0.j          0.

#### &#x270d; 13.4) Compute the sum of the lists X and Y without using the function for.

In [83]:
X = [1,2,4,5]
Y = [4,56,21,4]
list(np.array(X) + np.array(Y))

[5, 58, 25, 9]

#### &#x270d; 13.5) Compute the scalar product of X and Y.

Hint : the numpy function is `dot`.

A square matrix, in which all diagonal elements are equal to same scalar and all other elements are zero, is called a scalar matrix. Or. A diagonal matrix, in which all diagonal elements are equal to same scalar, is called a scalar matrix. A.

In [96]:
matX = np.arange(5,17).reshape(4,3)
matY = np.arange(21,27).reshape(3,2)
np.dot(matX, matY)
#matX.dot(matY)

array([[ 418,  436],
       [ 625,  652],
       [ 832,  868],
       [1039, 1084]])

## 14) Pandas

In Pandas, we work with dataframes. Let's create one to see how it works.

#### &#x270d; 14.1) Import the pandas module, with pd as shortened name, and the random module. 

In [253]:
import pandas as pd
import numpy as np
import random
import sys
import rstr
!{sys.executable} -m pip install rstr

#### &#x270d; 14.2) Create 5 lists composed of 4 random numbers between 1 and 100 and a string as final element.

Use random.randint(i,n) from numpy to get a list of n random numbers between 0 and i-1.

In [289]:
# A l'aide d'une fonction
def createList(n):
    L = []
    for i in range(n):
        L.append( list(np.random.randint(1,101,4)) + [rstr.xeger(r'[A-Z]{2,6}')])
    return L
print(createList(5))

# Solution
data_list = []

for i in range(5):
  random_4 = np.random.randint(low=1, high=100, size=4)
  a_or_b = np.random.choice(['a', 'b'])
  data_list.append(list(random_4) + [a_or_b])
print("\n",data_list)

# En une ligne à l'aide des listes en compréhension
[ list(np.random.randint(1,101,4)) + [rstr.xeger(r'[A-Z]{2,6}')] for i in range(5) ]

[[87, 28, 30, 39, 'CYHSJ'], [80, 29, 38, 94, 'RYOKS'], [23, 53, 43, 51, 'DXAKC'], [4, 53, 77, 76, 'LUESZD'], [42, 98, 36, 93, 'VKRTMO']]

 [[88, 46, 51, 26, 'b'], [96, 13, 73, 75, 'b'], [88, 45, 39, 93, 'a'], [74, 3, 41, 20, 'a'], [81, 81, 10, 51, 'a']]


[[30, 82, 20, 95, 'VI'],
 [7, 97, 14, 100, 'GDMDS'],
 [93, 64, 92, 33, 'QAJT'],
 [56, 69, 80, 83, 'QJW'],
 [97, 21, 16, 63, 'BZPIW']]

#### &#x270d; 14.3) Create a dataframe named data, composed of these 5 lists. We will call the columns F1, F2, F3, F4 and STR.
Reminder : use the function `pandas.DataFrame(list, columns = [...])` with `columns = ` the names of the columns in a list.



In [290]:
df = pd.DataFrame(createList(5), columns=['F1','F2','F3','F4','STR'])
df

F1  F2   F3  F4     STR
0  16  44   96  88  DTGPDK
1  74  34   34  71    DUYP
2  13  27   83  61      PW
3  90  89  100  25   QUFFN
4  91  28   16  74    IISW

#### &#x270d; 14.4) Print the STR feature of each element.

In [291]:
print(df['STR'])

0    DTGPDK
1      DUYP
2        PW
3     QUFFN
4      IISW
Name: STR, dtype: object


#### &#x270d; 14.5) Create a new dataframe composed of the rows 1, 3 and 4 of data. Compute its F3 mean.



In [293]:
df2 = df.iloc[[1,3,4],:] 
print(df2)
df2['F2'].mean()

   F1  F2   F3  F4    STR
1  74  34   34  71   DUYP
3  90  89  100  25  QUFFN
4  91  28   16  74   IISW


50.333333333333336

#### &#x270d; 14.6) Let's go back to the dataframe `data`. Create a new dataframe composed of the daqta rows where F1 > 40.



In [294]:
df

F1  F2   F3  F4     STR
0  16  44   96  88  DTGPDK
1  74  34   34  71    DUYP
2  13  27   83  61      PW
3  90  89  100  25   QUFFN
4  91  28   16  74    IISW

In [ ]:
df["F1"] > 40